dojo_plugin/api/v1/dojo.py

```diff
@@ -0,0 +1,244 @@
+
+
+@dojo_namespace.route("/delete")
+class DeleteDojo(Resource):
+    @authed_only
+    def post(self):
+        data = request.get_json()
+        user = get_current_user()
+
+        dojo_id = data.get("dojo_id")
+        dojo = Dojos.query.filter_by(id=dojo_id).first()
+        if not is_dojo_admin(user, dojo):
+            return {"success": False, "error": f"Invalid dojo specified: {data.get('dojo_id')}"}
+
+        dojo_dir = DOJOS_DIR/str(user.id)/dojo.id
+        if not dojo_dir.exists():
+            return {"success": False, "error": "Dojo directory does not exist."}
+
+        db.session.delete(dojo)
+        db.session.commit()
+
+        shutil.rmtree(str(dojo_dir))
+
+        return {"success": True, "dojo_id": dojo.id}
+
+@dojo_namespace.route("/create")
+class CreateDojo(Resource):
+    @authed_only
+    def post(self):
+        data = request.get_json()
+        user = get_current_user()
+
+        with tempfile.TemporaryDirectory(dir=DOJOS_DIR, prefix=str(user.id), suffix=".git-clone") as tmp_dir:
+            try:
+                dojo_repo = data.get("dojo_repo")
+                GIT_SSH_REGEX = "^git@github.com:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                GIT_HTTPS_REGEX = "^https://github.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$"
+                assert re.match(GIT_SSH_REGEX, dojo_repo) or re.match(GIT_HTTPS_REGEX, dojo_repo), (
+                    f"Repository violates regular expression. Must match <code>{GIT_SSH_REGEX}</code> or <code>{GIT_HTTPS_REGEX}</code>."
+                )
+
+                # clone it!
+                clone_dir = pathlib.Path(tmp_dir)/"clone"
+                clone_dir.mkdir()
+                returncode, output = sandboxed_git_clone(dojo_repo, str(ctfd_to_host_path(clone_dir)))
+
+                N=b"\n"
+                assert returncode == 0, (
+                    f"Dojo clone failed with error code {returncode}:<br>"
+                    f"<code>{output.replace(N,b'<br>').decode('latin1')}</code><br>"
+                    "Please make sure that you properly added the deploy key to the repository settings, "
+                    "and properly entered the repository URL."
+                )
+
+                # figure out the dojo ID
+                dojo_specs = list(clone_dir.glob("*.yml"))
+                assert len(dojo_specs) == 1, (
+                    f"Dojo repository must have exactly one top-level dojo spec yml named {{YOUR_DOJO_ID}}.yml. Yours has: {dojo_specs}"
+                )
+                dojo_id = dojo_specs[0].stem
+                assert id_regex(dojo_id), (
+                    f"Your dojo ID (the extensionless name of your .yml file) must be a valid URL component."
+                )
+
+                # make sure there aren't any symlink shenanigans
+                assert dojo_specs[0].is_file() and not dojo_specs[0].is_symlink(), (
+                    f"{dojo_specs[0].name} is not a regular file!"
+                )
```

Comment:
We can just open `dojo.yml` with O_NOFOLLOW

Reply:
Either way works for me, but then the pathlib integration isn't as magic...